File: 10 PrimeandComposite/Peaks/peaks.js
function solution(A) {
  if (A.length < 3) return 0;
  const peaks = [];

  for (let i = 1; i < A.length - 1; i += 1) {
    if (A[i] > A[i - 1] && A[i] > A[i + 1]) peaks.push(i);
  }
  if (peaks.length === 0) return 0;

  for (let i = peaks.length; i >= 2; i -= 1) {
    if (A.length % i === 0) {
      const K = A.length / i;
      let x = 0;
      for (const i in peaks) {
        const start = x * K;
        const end = (x + 1) * K;
        if (peaks[i] >= start && peaks[i] < end) { x += 1; }
      }
      if (x === i) return i;
    }
  }
  return 1;
}

//solution([1, 2, 3, 4, 3, 4, 1, 2, 3, 4, 6, 2]); // 3

File: 10 PrimeandComposite/Peaks/peaks.java
import java.util.ArrayList;

class Solution {
  public int solution(int[] A) {
    if (A.length < 3) return 0;
    ArrayList<Integer> peaks = new ArrayList<Integer>();
    for (int i = 1; i < A.length - 1; i += 1) {
      if (A[i] > A[i - 1] && A[i] > A[i + 1])
        peaks.add(i);
    }
    if (peaks.size() == 0) return 0;

    for (int i = peaks.size(); i >= 2; i -= 1) {
      if (A.length % i == 0) {
        int K = A.length / i;
        int x = 0;
        for (int y = 0; y < peaks.size(); y++) {
          int start = x * K;
          int end = (x + 1) * K;
          if (peaks.get(y) >= start && peaks.get(y) < end) {
            x += 1;
          }
        }
        if (x == i)
          return i;
      }
    }
    return 1;
  }
}
